File: Coupang/146.lru-cache.py
"""
Approach: Hash map plus doubly linked list.
Data structure: the map gives O(1) key lookup; the linked list gives O(1) removal and insertion for recency updates.
Interview logic: keep least recently used near the dummy head and most recently used near the dummy tail. get and existing-key put move a node to the tail; over-capacity put evicts head.next.
Complexity: O(1) average time per operation, O(capacity) space.
Tests and edge cases: capacity 1; updating an existing key must not evict; missing get returns -1.
"""
from __future__ import annotations

# @lc code=start
class _Node:
    def __init__(self, key: int = 0, value: int = 0):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None

class LRUCache:
    def __init__(self, capacity: int):
        self.capacity = capacity
        self.nodes = {}
        self.head = _Node()
        self.tail = _Node()
        self.head.next = self.tail
        self.tail.prev = self.head

    def _remove(self, node: _Node) -> None:
        node.prev.next = node.next
        node.next.prev = node.prev

    def _add_to_back(self, node: _Node) -> None:
        last = self.tail.prev
        last.next = node
        node.prev = last
        node.next = self.tail
        self.tail.prev = node

    def _mark_used(self, node: _Node) -> None:
        self._remove(node)
        self._add_to_back(node)

    def get(self, key: int) -> int:
        if key not in self.nodes:
            return -1
        node = self.nodes[key]
        self._mark_used(node)
        return node.value

    def put(self, key: int, value: int) -> None:
        if key in self.nodes:
            node = self.nodes[key]
            node.value = value
            self._mark_used(node)
            return
        node = _Node(key, value)
        self.nodes[key] = node
        self._add_to_back(node)
        if len(self.nodes) > self.capacity:
            lru = self.head.next
            self._remove(lru)
            del self.nodes[lru.key]
# @lc code=end

File: companies/Chewy/146.lru-cache.java
import java.util.*;

/**
 * Algorithm:
 * Combine a HashMap for key lookup with a doubly linked list for recency order.
 * The most recently used node sits before the tail sentinel; the least recently
 * used node sits after the head sentinel.
 *
 * Java data structures:
 * HashMap<Integer, Node> gives O(1) lookup. Custom linked nodes give O(1)
 * removal and insertion without relying on LinkedHashMap.
 *
 * Complexity:
 * get and put are O(1), space O(capacity).
 */
class LRUCache {
    private static class Entry {
        int key;
        int value;
        Entry prev;
        Entry next;

        Entry() {}

        Entry(int key, int value) {
            this.key = key;
            this.value = value;
        }
    }

    private final int capacity;
    private final Map<Integer, Entry> nodes;
    private final Entry head;
    private final Entry tail;

    public LRUCache(int capacity) {
        this.capacity = capacity;
        nodes = new HashMap<>();
        head = new Entry();
        tail = new Entry();
        head.next = tail;
        tail.prev = head;
    }

    public int get(int key) {
        Entry node = nodes.get(key);
        if (node == null) {
            return -1;
        }
        markUsed(node);
        return node.value;
    }

    public void put(int key, int value) {
        Entry node = nodes.get(key);
        if (node != null) {
            node.value = value;
            markUsed(node);
            return;
        }
        node = new Entry(key, value);
        nodes.put(key, node);
        addToBack(node);
        if (nodes.size() > capacity) {
            Entry lru = head.next;
            remove(lru);
            nodes.remove(lru.key);
        }
    }

    private void markUsed(Entry node) {
        remove(node);
        addToBack(node);
    }

    private void remove(Entry node) {
        node.prev.next = node.next;
        node.next.prev = node.prev;
    }

    private void addToBack(Entry node) {
        Entry last = tail.prev;
        last.next = node;
        node.prev = last;
        node.next = tail;
        tail.prev = node;
    }
}

